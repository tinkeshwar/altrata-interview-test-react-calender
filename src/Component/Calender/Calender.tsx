import classNames from "classnames"
import { forwardRef } from "react"

interface CalenderProps {
  className?: string
  date: Date
}

interface DayObjectInterface {
  day: number
  narrow: string
  short: string
  long: string
}

interface DaysObjectInterface {
  type: string
  date: number
}

export const Calender = forwardRef<HTMLDivElement, CalenderProps>((
  {
    date, 
    className
  },
  ref
) => {

  // To check whether valid date object is recieved or not.
  if(Object.prototype.toString.call(date) !== "[object Date]"){
    console.error('Not a valid date object passed to calender component.')
    return <></>
  }

  const _classname: string = classNames(
    'custom-calender-class', // to customise component globally
    className
  )

  const month: string = date.toLocaleString('default', { month: 'long' })
  const year: number = date.getFullYear()

  return (
    <div ref={ref} className={_classname}>
      <div className={'row border'}>
        <div className={'col-12 text-center border-bottom fs-3'}>{month} {year}</div>
        <div className={'row gx-0'}>
          <GetMonthHeader/>
        </div>
        <GetWeekRows date={date}/>
      </div>
    </div>
  )
})


// generate header for calender.
const GetMonthHeader = ({header='narrow'}:{header?: 'narrow'|'short'|'long'}) => {
  const dayObject: DayObjectInterface[]  = [
    {day: 1, narrow: 'Su', short: 'Sun', long: 'Sunday'},
    {day: 2, narrow: 'Mo', short: 'Mon', long: 'Monday'},
    {day: 3, narrow: 'Tu', short: 'Tue', long: 'Tuesday'},
    {day: 4, narrow: 'We', short: 'Wed', long: 'Wednesday'},
    {day: 5, narrow: 'Th', short: 'Thu', long: 'Thursday'},
    {day: 6, narrow: 'Fr', short: 'Fri', long: 'Friday'},
    {day: 7, narrow: 'Sa', short: 'Sat', long: 'Saturday'},
  ]
  return <>{dayObject && dayObject.map((day: DayObjectInterface, index: number) => <div key={`days-${index}`} className={'text-center bg-dark text-white col border fw-bold'}>{day[header]}</div>)}</>
}

// generate calender rows
const GetWeekRows = ({date}:{date: Date}) => {
  const day: string[] = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']
  const today: number = Number(new Date(date).toLocaleDateString('default', { day: 'numeric'}))
  const firstDay: number = day.indexOf(new Date(date.getFullYear(), date.getMonth(), 1).toLocaleString('default', { weekday: 'short' }))+1
  const lastDay: number = Number(new Date(date.getFullYear(), date.getMonth()+1, 0).toLocaleString('default', { day: 'numeric' }))
  const lastMonthLastDay: number = Number(new Date(date.getFullYear(), date.getMonth(), 0).toLocaleString('default', { day: 'numeric' }))
  const weeks: number[] = [1, 2, 3, 4, 5, 6]

  // generate days array for a month
  const days = (): DaysObjectInterface[] => {
    const total: DaysObjectInterface[] = []
    for(let i = (lastMonthLastDay-(firstDay-2)); i<=lastMonthLastDay; i++) total.push({type: 'past', date: i})
    for(let i = 1; i<=lastDay; i++) total.push({type: 'current', date: i})
    const totalLength= total.length
    for(let i = 1; i<=42-totalLength; i++) total.push({type: 'future', date: i})
    return total
  }
  
  return <>{weeks.map((week: number, index: number) => <div key={`week-${index}`} className={'row gx-0'}><GetWeekRow week={week} days={days()} today={today}/></div>)}</>
}


// generate calender row
const GetWeekRow = ({week, days, today}: {week: number, days: DaysObjectInterface[], today: number}) => {
  const weekDays = (): DaysObjectInterface[] => {
    return days.slice((7*(week-1)), 7*(week))
  }

  const classname = (week: number, day: DaysObjectInterface, today: number): string =>{
    return classNames(
      'text-center',
      'col',
      'border',
      (['past', 'future'].includes(day.type)) && 'bg-light text-white',
      (day.date === today && day.type === 'current') && 'bg-primary text-white'
    )
  }

  return <> {weekDays().map((day: DaysObjectInterface, index: number) => <div key={`day-${index}`} className={classname(week, day, today)}>{day.date}</div>)}</>
}