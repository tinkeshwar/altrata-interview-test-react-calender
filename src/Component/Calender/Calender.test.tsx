import { render } from '@testing-library/react';
import { Calender } from './Calender';

test('Render calender', () => {
  const { container } = render(<Calender date={new Date('01/01/2023')} className={'test-case-calender'}/>)
  expect(container).toMatchSnapshot()
})

test('Class name prop', () => {
  const { container } = render(<Calender date={new Date()} className={'test-case-calender'}/>)
  expect(container.firstElementChild).toHaveClass('test-case-calender')
})