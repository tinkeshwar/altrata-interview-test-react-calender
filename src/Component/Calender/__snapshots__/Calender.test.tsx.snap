// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render calender 1`] = `
<div>
  <div
    class="custom-calender-class test-case-calender"
  >
    <div
      class="row border"
    >
      <div
        class="col-12 text-center border-bottom fs-3"
      >
        January
         
        2023
      </div>
      <div
        class="row gx-0"
      >
        <div
          class="text-center bg-dark text-white col border fw-bold"
        >
          Su
        </div>
        <div
          class="text-center bg-dark text-white col border fw-bold"
        >
          Mo
        </div>
        <div
          class="text-center bg-dark text-white col border fw-bold"
        >
          Tu
        </div>
        <div
          class="text-center bg-dark text-white col border fw-bold"
        >
          We
        </div>
        <div
          class="text-center bg-dark text-white col border fw-bold"
        >
          Th
        </div>
        <div
          class="text-center bg-dark text-white col border fw-bold"
        >
          Fr
        </div>
        <div
          class="text-center bg-dark text-white col border fw-bold"
        >
          Sa
        </div>
      </div>
      <div
        class="row gx-0"
      >
         
        <div
          class="text-center col border bg-primary text-white"
        >
          1
        </div>
        <div
          class="text-center col border"
        >
          2
        </div>
        <div
          class="text-center col border"
        >
          3
        </div>
        <div
          class="text-center col border"
        >
          4
        </div>
        <div
          class="text-center col border"
        >
          5
        </div>
        <div
          class="text-center col border"
        >
          6
        </div>
        <div
          class="text-center col border"
        >
          7
        </div>
      </div>
      <div
        class="row gx-0"
      >
         
        <div
          class="text-center col border"
        >
          8
        </div>
        <div
          class="text-center col border"
        >
          9
        </div>
        <div
          class="text-center col border"
        >
          10
        </div>
        <div
          class="text-center col border"
        >
          11
        </div>
        <div
          class="text-center col border"
        >
          12
        </div>
        <div
          class="text-center col border"
        >
          13
        </div>
        <div
          class="text-center col border"
        >
          14
        </div>
      </div>
      <div
        class="row gx-0"
      >
         
        <div
          class="text-center col border"
        >
          15
        </div>
        <div
          class="text-center col border"
        >
          16
        </div>
        <div
          class="text-center col border"
        >
          17
        </div>
        <div
          class="text-center col border"
        >
          18
        </div>
        <div
          class="text-center col border"
        >
          19
        </div>
        <div
          class="text-center col border"
        >
          20
        </div>
        <div
          class="text-center col border"
        >
          21
        </div>
      </div>
      <div
        class="row gx-0"
      >
         
        <div
          class="text-center col border"
        >
          22
        </div>
        <div
          class="text-center col border"
        >
          23
        </div>
        <div
          class="text-center col border"
        >
          24
        </div>
        <div
          class="text-center col border"
        >
          25
        </div>
        <div
          class="text-center col border"
        >
          26
        </div>
        <div
          class="text-center col border"
        >
          27
        </div>
        <div
          class="text-center col border"
        >
          28
        </div>
      </div>
      <div
        class="row gx-0"
      >
         
        <div
          class="text-center col border"
        >
          29
        </div>
        <div
          class="text-center col border"
        >
          30
        </div>
        <div
          class="text-center col border"
        >
          31
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          1
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          2
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          3
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          4
        </div>
      </div>
      <div
        class="row gx-0"
      >
         
        <div
          class="text-center col border bg-light text-white"
        >
          5
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          6
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          7
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          8
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          9
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          10
        </div>
        <div
          class="text-center col border bg-light text-white"
        >
          11
        </div>
      </div>
    </div>
  </div>
</div>
`;
